modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/ResolverSupplier.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.metadata;
+
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.mule.runtime.api.util.Preconditions.checkArgument;
+import static org.mule.runtime.core.util.ClassUtils.getClassName;
+
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.core.internal.metadata.NullMetadataResolverSupplier;
+import org.mule.runtime.core.util.ClassUtils;
+import org.mule.runtime.extension.api.metadata.NullMetadataResolver;
+
+import java.util.function.Supplier;
+
+/**
+ * {@link Supplier} implementation that returns instances of type {@link T} from a given {@link Class}
+ *
+ * @since 4.0
+ * @param <T> The type of the instances to supply
+ */
+public final class ResolverSupplier<T> implements Supplier<T> {
```

Comment:
if this is a resolver supplier, then T should extend Resolver. Otherwise, this is just a supplier.